rusk/src/lib/lib.rs

```diff
@@ -158,6 +177,26 @@ impl Rusk {
                 },
             };
 
+            // If the gas spent is larger that the remaining block gas, then we
+            // re-execute the transactions with a new session. We can ignore the
+            // results since we know they will be valid.
+            if gas_spent > block_gas_left {
+                session = rusk_abi::new_session(
+                    &inner.vm,
+                    current_commit,
+                    block_height,
+                )?;
+
+                for spent_tx in &spent_txs {
+                    let _: (
+                        u64,
+                        Option<Result<RawResult, ContractError>>,
+                    ) =
+                    session.call(TRANSFER_CONTRACT, "execute", &spent_tx.0)
+                        .expect("Re-execution of spent transactions should never fail");
+                }
+            }
+
```

Comment:
If the gas spent is greater than the gas available, we should not include the tx in the block.

If I understood properly, you're resetting the session to the previous checkpoint, right? If so, is a `continue` missing at the end of the scope?

Reply:
It is, will change.